hw1/task1/get_stats.py

```diff
@@ -0,0 +1,169 @@
+
+    gen_stat['longest_ride'] = longest
+    gen_stat['mean_cost'] = clean_data['total_amount'].mean()
+
+    pickup_times = pd.DataFrame(clean_data['lpep_pickup_datetime'])
+    if len(pickup_times) != 0:
+        pickup_times = pickup_times.reindex(pickup_times['lpep_pickup_datetime'])
+        resampled_data = pickup_times['lpep_pickup_datetime'].resample('10T')
+        ten_minutes_resample = resampled_data.count()
+
+        max_count_start = ten_minutes_resample.idxmax()
+        max_count = ten_minutes_resample.max()
+        max_count_end = max_count_start + np.timedelta64(10, 'm')
+
+        gen_stat['max_count'] = max_count
+        gen_stat['max_count_start'] = max_count_start
+        gen_stat['max_count_end'] = max_count_end
+    else:
+        gen_stat['max_count'] = None
+        gen_stat['max_count_start'] = None
+        gen_stat['max_count_end'] = None
+
+    gen_stat['invalid_rows'] = invalid_rows
+
+    if return_count:
+        gen_stat['count'] = clean_data.shape[0]
+    gen_stat = pd.DataFrame(gen_stat, index=[0])
+    return gen_stat
+
+
+def missing_dates(data):
+    clean_data = data.dropna().copy()
+
+    max_time = clean_data['lpep_pickup_datetime'].max()
+    min_time = clean_data['lpep_pickup_datetime'].min()
+
+    if max_time is pd.NaT or min_time is pd.NaT:
+        date_range = None
+    else:
+        date_range = pd.date_range(min_time, max_time, freq='D').date
+    missing_date = pd.Series(date_range)
+    missing_date.name = 'missing_date'
+
+    valid_dates = clean_data['lpep_pickup_datetime'].apply(lambda i: i.date())
+
+    missing_date = pd.DataFrame(missing_date.apply(lambda s: None if s in valid_dates.values else s).dropna())
+    return missing_date
+
+
+def usage_stat(data):
+    clean_data = data.dropna().copy()
+    stat = clean_data.groupby(clean_data['lpep_pickup_datetime'].dt.date)['lpep_pickup_datetime'].count()
+    stat = pd.DataFrame(stat)
+    stat.index.name = 'date'
+    stat.columns = ['count']
+    return stat
+
+
+def trip_stat(data):
+    clean_data = data.dropna().copy()
+    clean_data['trip_duration'] = trip_durations(clean_data['lpep_pickup_datetime'],
+                                                 clean_data['lpep_dropoff_datetime'])
+
+    clean_data.loc[:, 'passenger_count'] = clean_data['passenger_count'].astype(int)
+    clean_data.loc[:, 'month'] = clean_data['lpep_pickup_datetime'].apply(lambda d: d.month)
+    clean_data = clean_data.dropna()
+
+    if len(clean_data['lpep_pickup_datetime']) != 0:
+        trip_stat = pd.DataFrame(clean_data.groupby(['month', 'passenger_count'])['trip_duration'].mean())
+        trip_stat.loc[:, 'trip_duration'] = pd.to_timedelta(trip_stat['trip_duration'].apply(int), unit='s').apply(get_time)
+    else:
+        trip_stat = pd.DataFrame(columns=['month', 'passenger_count'])
+
+    return trip_stat
+
+
+def trip_durations(start, end):
+    trip_duration = end - start
+    trip_duration = trip_duration.dt.total_seconds()
```

Comment:
wile file that has only one row with wrong format of date got the following error:
```/home/maksim/Documents/work/big-data-lab/2020/green-taxi/TLC_trip_data/venv/lib/python3.7/site-packages/pandas/core/arrays/datetimelike.py:1215: PerformanceWarning: Adding/subtracting array of DateOffsets to DatetimeArray not vectorized
  PerformanceWarning,
Traceback (most recent call last):
  File "/home/maksim/Documents/work/big-data-lab/2020/green-taxi/TLC_trip_data/hw1/task1/main.py", line 15, in <module>
    _main()
  File "/home/maksim/Documents/work/big-data-lab/2020/green-taxi/TLC_trip_data/hw1/task1/main.py", line 8, in _main
    stat.general_stats(data, return_count=False).to_csv('./stats/gen_stat.csv', index=False)
  File "/home/maksim/Documents/work/big-data-lab/2020/green-taxi/TLC_trip_data/hw1/task1/get_stats.py", line 54, in general_stats
    clean_data['lpep_dropoff_datetime'])
  File "/home/maksim/Documents/work/big-data-lab/2020/green-taxi/TLC_trip_data/hw1/task1/get_stats.py", line 141, in trip_durations
    trip_duration = trip_duration.dt.total_seconds()
AttributeError: 'DatetimeProperties' object has no attribute 'total_seconds'```